File: tests/examples/java/simple.java
public class simple {
    public static void main(String[] args) {
        // Initialize counters for assignments and evaluations
        int assignments = 0;
        int evaluations = 0;

        // Start measuring time and memory
        long startTime = System.nanoTime();
        Runtime runtime = Runtime.getRuntime();
        runtime.gc(); // Suggest garbage collection to minimize memory usage at the start
        long startMemory = runtime.totalMemory() - runtime.freeMemory();

        // Initialize variables
        int x = 10;
        int y = 20;
        assignments += 2; // Initial assignments of x and y

        // Number of iterations
        int n = 1000;
        int i = 0;
        assignments += 2; // Assigning n and i

        // Perform the loop for accumulation
        while (i < n) {
            // Assign new values by adding two-digit numbers
            x += 15;
            assignments += 1; // Assignment of x += 15
            evaluations += 1; // Evaluation of x + 15
            System.out.println("Iteration " + (i + 1) + " - Step 1: x = " + x);

            x += 20;
            assignments += 1; // Assignment of x += 20
            evaluations += 1; // Evaluation of x + 20
            System.out.println("Iteration " + (i + 1) + " - Step 2: x = " + x);

            y += 25;
            assignments += 1; // Assignment of y += 25
            evaluations += 1; // Evaluation of y + 25
            System.out.println("Iteration " + (i + 1) + " - Step 1: y = " + y);

            y += 30;
            assignments += 1; // Assignment of y += 20
            evaluations += 1; // Evaluation of y + 20
            System.out.println("Iteration " + (i + 1) + " - Step 2: y = " + y);

            // Increment the iterator
            i++;
            assignments += 1; // Assignment of i += 1
            evaluations += 1; // Evaluation of i + 1
        }

        // Reset the iterator
        i = 0;
        assignments += 1; // Reassigning i to 0

        // Perform the loop for reversal
        while (i < n) {
            // Reverse the state to the previous value
            x -= 15;
            assignments += 1; // Assignment of x -= 20
            evaluations += 1; // Evaluation of x - 20
            System.out.println("Reverse Iteration " + (i + 1) + " - Step 1: x = " + x);

            x -= 20;
            assignments += 1; // Assignment of x -= 15
            evaluations += 1; // Evaluation of x - 15
            System.out.println("Reverse Iteration " + (i + 1) + " - Step 2: x = " + x);

            y -= 25;
            assignments += 1; // Assignment of y -= 20
            evaluations += 1; // Evaluation of y - 20
            System.out.println("Reverse Iteration " + (i + 1) + " - Step 1: y = " + y);

            y -= 30;
            assignments += 1; // Assignment of y -= 25
            evaluations += 1; // Evaluation of y - 25
            System.out.println("Reverse Iteration " + (i + 1) + " - Step 2: y = " + y);

            // Increment the iterator
            i++;
            assignments += 1; // Assignment of i += 1
            evaluations += 1; // Evaluation of i + 1
        }

        // Stop measuring time and memory
        long endTime = System.nanoTime();
        long endMemory = runtime.totalMemory() - runtime.freeMemory();

        // Calculate execution time in seconds
        double executionTime = (endTime - startTime) / 1_000_000_000.0;

        // Calculate memory usage in megabytes
        double memoryUsed = (endMemory - startMemory) / (1024.0 * 1024.0);

        // Print final values of x and y
        System.out.println("Final x: " + x);
        System.out.println("Final y: " + y);

        // Display the execution time, memory usage, assignments, and evaluations
        System.out.println("\nExecution Time: " + executionTime + " seconds");
        System.out.println("Memory Used: " + memoryUsed + " MB");
        System.out.println("Total Assignments: " + assignments);
        System.out.println("Total Evaluations: " + evaluations);
    }
}

File: tests/examples/rust/simple.rs
use cfg_if::cfg_if;
use sysinfo::{Pid, ProcessExt, System, SystemExt};
use std::time::Instant;

fn main() {
    let mut assignments = 0;
    let mut evaluations = 0;

    let start_time = Instant::now();
    let start_memory = get_memory_usage();

    let mut x = 10;
    let mut y = 20;
    assignments += 2;

    let n = 1000;
    let mut i = 0;
    assignments += 2;

    while i < n {
        x += 15;
        assignments += 1;
        evaluations += 1;
        println!("Iteration {} - Step 1: x = {}", i + 1, x);

        x += 20;
        assignments += 1;
        evaluations += 1;
        println!("Iteration {} - Step 2: x = {}", i + 1, x);

        y += 25;
        assignments += 1;
        evaluations += 1;
        println!("Iteration {} - Step 1: y = {}", i + 1, y);

        y += 30;
        assignments += 1;
        evaluations += 1;
        println!("Iteration {} - Step 2: y = {}", i + 1, y);

        i += 1;
        assignments += 1;
        evaluations += 1;
    }

    i = 0;
    assignments += 1;

    while i < n {
        // Check to prevent underflow
        if x >= 15 {
            x -= 15;
            assignments += 1;
            evaluations += 1;
            println!("Reverse Iteration {} - Step 1: x = {}", i + 1, x);
        } else {
            println!("Reverse Iteration {} - Step 1: x cannot be reduced further", i + 1);
            break;
        }

        if x >= 20 {
            x -= 20;
            assignments += 1;
            evaluations += 1;
            println!("Reverse Iteration {} - Step 2: x = {}", i + 1, x);
        } else {
            println!("Reverse Iteration {} - Step 2: x cannot be reduced further", i + 1);
            break;
        }

        if y >= 25 {
            y -= 25;
            assignments += 1;
            evaluations += 1;
            println!("Reverse Iteration {} - Step 1: y = {}", i + 1, y);
        } else {
            println!("Reverse Iteration {} - Step 1: y cannot be reduced further", i + 1);
            break;
        }

        if y >= 30 {
            y -= 30;
            assignments += 1;
            evaluations += 1;
            println!("Reverse Iteration {} - Step 2: y = {}", i + 1, y);
        } else {
            println!("Reverse Iteration {} - Step 2: y cannot be reduced further", i + 1);
            break;
        }

        i += 1;
        assignments += 1;
        evaluations += 1;
    }

    let end_time = Instant::now();
    let end_memory = get_memory_usage();

    let execution_time = end_time.duration_since(start_time);
    let memory_used = (end_memory - start_memory) as f64 / (1024.0 * 1024.0);

    println!("Final x: {}", x);
    println!("Final y: {}", y);
    println!("\nExecution Time: {:.6} seconds", execution_time.as_secs_f64());
    println!("Memory Used: {:.6} MB", memory_used);
    println!("Total Assignments: {}", assignments);
    println!("Total Evaluations: {}", evaluations);
}

fn get_memory_usage() -> usize {
    let mut system = System::new_all();
    system.refresh_all();

    let pid = Pid::from(std::process::id() as usize);
    if let Some(process) = system.process(pid) {
        process.memory() as usize
    } else {
        0
    }
}
